File: src/Python/29.DivideTwoIntegers.py
class Solution:
    def divide(self, dividend, divisor):
        """
        :type dividend: int
        :type divisor: int
        :rtype: int
        """
        positive = (dividend > 0) == (divisor > 0) # 检测是否同为正负
        dividend, divisor = abs(dividend), abs(divisor) 
        rlt = 0
        while dividend >= divisor: # 如果除数太大了，就重新开始
            temp_divisor, i = divisor, 1 # 更新除数和计量单位
            while dividend >= temp_divisor:
                dividend -= temp_divisor
                rlt += i
                i = i<<1
                temp_divisor = temp_divisor << 1
        if not positive:
            rlt = -rlt
        return min(max(rlt, -2**31), 2**31-1) # 返回限制内的数字
            

if __name__ == "__main__":
    solution = Solution()
    print(solution.divide(10, 3))
    print(solution.divide(7, -3))
    print(solution.divide(1, -1))
    print(solution.divide(0, 1))
    print(solution.divide(-1, 1))
    print(solution.divide(-2147483648, -1))

File: src/Java/1-100/29.DivideTwoIntegers.java

import java.awt.List;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.Comparator;
import java.util.HashMap;
import java.util.Queue;
import java.util.HashSet;
import java.util.LinkedList;


class Solution {
    public int divide(int dividend, int divisor) {
        if (divisor == 0)
        return 0;
    int max = Integer.MAX_VALUE;
    int min = Integer.MIN_VALUE;
    long divid = (long)dividend;
    long divi = (long)divisor;
    if (divi == 1)
        return (int)divid;
    if (divi == -1){
        if (divid <= min)
            return max;
        return (int)-divid;
    }
    boolean flag = true;
    if (divid < 0){
        divid = -divid;
        if (divi < 0){
            divi = -divi;
        }
        else{
            flag = false;
        }
    }
    else if (divi < 0){
        divi = -divi;
        flag = false;
    }
    long res = 0;
    long tmp = 0;
    long cnt = 1;
    while (divi <= divid){
        cnt = 1;
        tmp = divi;
        while (tmp <= divid){
            tmp <<= 1;
            cnt <<= 1;
        }
        res += (cnt >> 1);
        divid -= (tmp >> 1);
    }
    return flag ? (int)res : (int)-res;
    }
}
